### DISCUSSION TITLE
Send termination request

### DISCUSSION THREAD

**Discussion by OskarKlintrot** [adamralph/simple-exec#405] [Ideas]
I couldn't figure out a way to gracefully shutdown a dotnet app I'm testing. I wish I could send a termination request to let the process know that I want it to try to shut down and still be able to read the output from it. Something like this:

```csharp
var cancellationTokenSource = new CancellationTokenSource(TimeSpan.FromMinutes(2));
var terminationTokenSource = new CancellationTokenSource(); // Maybe not a cts but something 
                                                            // similar that better indicates
                                                            // that it's the process that should 
                                                            // shut down and not the task 
                                                            // that should be cancelled

var webApiTask = ReadAsync(
    "dotnet",
    "run",
    workingDirectory: "path-to-my-api",
    terminationRequest: terminationTokenSource.Token,
    cancellationToken: cancellationTokenSource.Token);

// Ping webapi

terminationTokenSource.Cancel();

var (output, error) = await webApiTask; // Not cancelled unless it has passed more than 2 minutes

// Assert output and error
```

I can use the `cancellationToken` to at least kill the process and move on but I wish there were a way to still capture the output.

**Comment by adamralph** [comment]
Regarding determining whether the process was cancelled or whether it timed out, this is the way to do it:

```c#
var timeoutTokenSource = new CancellationTokenSource(TimeSpan.FromMinutes(2));
var terminationTokenSource = new CancellationTokenSource();

var combinedTokenSource =
    CancellationTokenSource.CreateLinkedTokenSource(timeoutTokenSource.Token, terminationTokenSource.Token);

var webApiTask = ReadAsync(
    "dotnet",
    "run",
    workingDirectory: "path-to-my-api",
    cancellationToken: combinedTokenSource.Token);

// Ping webapi

terminationTokenSource.Cancel();

try
{
    var (output, error) = await webApiTask; // Not cancelled unless it has passed more than 2 minutes
}
catch (OperationCanceledException ex)) when (terminationTokenSource.IsCancellationRequested)
{
    // the process was terminated
}
```

However, this won't let you capture the output or error streams. While we could try and throw a custom exception, inheriting from `OperationCancelledException`, with `Output` and `Error` properties, I'm not sure it will work, given the [simple approach taken by SimpleExec](https://github.com/adamralph/simple-exec/blob/a8053da57e1e785d9e776d825a1fafac71a75acf/SimpleExec/Command.cs#L194-L208) to reading those streams.

Taking a step back for a moment, I think your use case may be a bit further advanced than the simple use cases SimpleExec is designed for. Perhaps you should look into using a more advanced package such as [CliWrap](https://github.com/Tyrrrz/CliWrap).

**Reply by OskarKlintrot** [reply]
I have another implementation I've used before that reads via events instead and that pipes the event back out to the user; https://gist.github.com/OskarKlintrot/3947da8d31c4fa275525e4afc4138e28#file-processrunner-cs-L17-L18. That way I have the output when the process is cancelled. Not that much more complicated ;)

Yeah, I might have to change to CliWrap even though I found your API much nicer to work with!
